src/daemon/distribution.c

```diff
@@ -0,0 +1,148 @@
+#include <math.h>
+#include "collectd.h"
+#include "distribution.h"
+
+struct bucket_s{
+	size_t bucket_counter;
+	double min_boundary;
+	double max_boundary;
+};
+
+struct distribution_s {
+     bucket_t *buckets;
+     size_t num_buckets;
+     size_t total_scalar_count;
+     double raw_data_sum;
+};
+
+//bucket constructor
+bucket_t initialize_bucket(double min_boundary, double max_boundary) {
+    bucket_t new_bucket = {
+        .bucket_counter = 0,
+        .min_boundary = min_boundary,
+        .max_boundary = max_boundary,
+    };
+    return new_bucket;
+} 
+
+//increment bucket_counter of specific bucket
+void increment(bucket_t *bucket) { 
+    bucket->bucket_counter++;
+} 
+
+distribution_t * distribution_new_linear(size_t num_buckets, double size) {
```

Comment:
It's good practice to check the arguments in public functions. E.g.:

```c
if ((num_buckets == 0) || (size <= 0)) {
  errno = EINVAL;
  return NULL;
}
```

Here and in the other constructor functions.

Reply:
Thank you, I added the checks.